### PR TITLE
fix: improve typing of property error field

### PR DESCRIPTION
**Pull request by Averethel**
References https://autoricardo.atlassian.net/browse/DM-788

## Motivation and context

During the migration to `react-hook-form` we discovered an error in typing the `fieldErrors` in the error response.

## Before

`fieldErrors` were typed as `keyof Data` (where `Data` is the type of API request data), which worked perfectly fine but only if we consider flat objects. Since the API request can contain a nested object this type was not specific enough.

i.e. if your data was a `MessageLead`

```typescript
type MessageLead = {
  email: string;
  interestedInFinancing: boolean;
  interestedInInsurance: boolean;
  interestedInLeasing: boolean;
  language: Language;
  message: string;
  firstName: string;
  lastName: string;
  phone: string;
  hasTestDrive?: boolean;
  testDrive?: { date?: string };
};
```

`fieldError.property` was typed as:

```typescript
type FieldError<MessageLead>.property =
  | 'email'
  | 'interestedInFinancing'
  | 'interestedInInsurance'
  | 'interestedInLeasing'
  | 'language'
  | 'message'
  | 'firstName'
  | 'lastName'
  | 'phone'
  | 'hasTestDrive'
  | 'testDrive';
```

notably `testDrive.date` is not a valid property.

## After

To remedy that a `Path` utility type was introduced. This new type maps nested object keys and concatenates them with the parent key. This ensures correct typing of `fieldError.property` and allows it to be used with the `react-hook-form` error handling

### CODE REVIEW

**Review by lkappeler — COMMENTED** [review]
I think it's looking fine, what I am wondering is where the actual errors go. Can yu add an example request and response body that will be typed by this

File: src/responseType.ts [hunk 1]
```diff
@@ -3,9 +3,17 @@ type BaseError = {
   message: string;
 };
 
-type FieldError<RequestData> = BaseError & { property: keyof RequestData };
+type Path<ObjectType extends object> = {
```
Comment (talamcol) [inline src/responseType.ts:6]:
```suggestion
type Path<ObjectType extends Record<string, unknown>> = {
```
I think it's discouraged to use the `object` type

File: src/responseType.ts [hunk 2]
```diff
@@ -3,9 +3,17 @@ type BaseError = {
...
+  [Key in keyof ObjectType & (string | number)]: ObjectType[Key] extends object
+    ? `${Key}` | `${Key}.${Path<ObjectType[Key]>}`
+    : `${Key}`;
+}[keyof ObjectType & (string | number)];
```
Comment (talamcol) [inline src/responseType.ts:10]: I was wondering, if you do: 
```typescript
type Foo = Path<{
  email: string;
  testDrive?: { date?: string };
}>;
// -> "testDrive" | "testDrive.date" | "email"
```
do you want the testDrive or only the testDrive.date one?

**Review by talamcol — APPROVED** [review]

File: src/responseType.ts (hunk 1 above)
Comment (Averethel) [reply src/responseType.ts:6]: I need to check if it still works with Record. Somehow I am not getting any linter/type-checker warnings about this anymore. I'll check if we disabled the rule or if things have changed.

File: src/responseType.ts (hunk 2 above)
Comment (Averethel) [reply src/responseType.ts:10]: I'm interested in both. The reason is - we can have a specific error on nested object fields or a more generic one on the entire nested object. I'd like to support both cases

Comment (talamcol) [reply src/responseType.ts:10]: ah got it, nicely solved 👍

**Comment by Averethel** [comment]
> I think it's looking fine, what I am wondering is where the actual errors go. Can you add an example request and response body that will be typed by this

Sure thing. Let's consider:
```typescript
type MessageLead = {
  email: string;
  interestedInFinancing: boolean;
  interestedInInsurance: boolean;
  interestedInLeasing: boolean;
  language: Language;
  message: string;
  firstName: string;
  lastName: string;
  phone: string;
  hasTestDrive?: boolean;
  testDrive?: { date?: string };
};

```

When we get the following validation response:
```json
{
  "code": "VALIDATION_FAILED",
  "fieldErrors": [
    {
      "code": "REQUIRED_NOT_NULL",
      "property": "testDrive.date"
    },
    {
      "code": "REQUIRED_NOT_BLANK",
      "property": "lastName"
    }
  ]
}
```

**Before**
`property` was typed as: 
```typescript
  "email" | 
  "interestedInFinancing" | 
  "interestedInInsurance" | 
  "interestedInLeasing" | 
  "language" |
  "message" | 
  "firstName" | 
  "lastName" | 
  "phone" | 
  "hasTestDrive" | 
  "testDrive"
```

***After**
```typescript
  "email" | 
  "interestedInFinancing" | 
  "interestedInInsurance" | 
  "interestedInLeasing" | 
  "language" |
  "message" | 
  "firstName" | 
  "lastName" | 
  "phone" | 
  "hasTestDrive" | 
  "testDrive" |
  "testDrive.date" 
```

Before the change the type system doesn't see error on the `testDrive.date` correctly

File: src/responseType.ts (hunk 1 above)
Comment (Averethel) [reply src/responseType.ts]: It turns out that `object` is correct here and `Record<string, unknown>` is causing some issues (too deep type declaration).

**Review by lkappeler — APPROVED** [review]
Thanks for the explanations. 

(In case someone wonders about the object type https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type)

**Pull request merged**